File: cpp/rotation_mpcd.cpp
#include "parameters.hpp"
#include <assert.h>

void rotation_mpcd() {
	int j, k, **cell_part, cell_no, *fluid_no;
	double cell_velx[lx*ly*lz], cell_vely[lx*ly*lz], cell_velz[lx*ly*lz], r[4], theta, phi, rho, rot[4][4], rr[4];
	double del_vx, del_vy, del_vz, *tmp_pos, var, del_vx1[no_of_fluid], del_vy1[no_of_fluid], del_vz1[no_of_fluid];
	double scale_fac_mpcd;

	tmp_pos = (double *)malloc(sizeof(double)*(3*no_of_fluid + 2));
	fluid_no = (int *)malloc(sizeof(int)*(lx*ly*lz + 1));
	cell_part = (int **)malloc(sizeof(int *)*(maxpart + 2));
	memset(fluid_no, 0, (lx*ly*lz + 2)*sizeof(int));
	std::copy(pos_fl, pos_fl + 3*no_of_fluid + 2, tmp_pos);

	for(int i = 0; i <= maxpart; i++)
		cell_part[i] = (int *)malloc(sizeof(int)*(lx*ly*lz + 2));

	for(int i = 1; i <= 3; i++)
		rr[i] = ran() - 0.5;

	for(int i=1; i <= no_of_fluid; i++) {
		tmp_pos[3*i-2] = mod(tmp_pos[3*i - 2] + rr[1], lx);
		tmp_pos[3*i-1] = mod(tmp_pos[3*i-1] + rr[2], ly);
		tmp_pos[3*i] = mod(tmp_pos[3*i] + rr[3], lz);
	}
	for(int i=1; i <= no_of_fluid; i++) {
		cell_no = 1 + int(tmp_pos[3*i-2]) + lx*int(tmp_pos[3*i-1]) + lx*ly*int(tmp_pos[3*i]);
		fluid_no[cell_no] = fluid_no[cell_no] + 1;
		j = fluid_no[cell_no];
		cell_part[j][cell_no] = i;
	}
	memset(cell_velx, 0, sizeof cell_velx);
	memset(cell_vely, 0, sizeof cell_vely);
	memset(cell_velz, 0, sizeof cell_velz);
	for(int i = 1; i <= lx*ly*lz; i++) {
		if (fluid_no[i] > 1) {
			for(int j = 1; j <= fluid_no[i]; j++) {
				k = cell_part[j][i];
				cell_velx[i] += vel_fl[3*k-2]/fluid_no[i];
				cell_vely[i] += vel_fl[3*k-1]/fluid_no[i];
				cell_velz[i] += vel_fl[3*k]/fluid_no[i];
			}
			rho = 2*ran() - 1, phi = 4.0*asin(1.0)*ran();
			r[1] = cos(phi)*sqrt(1-rho*rho);
			r[2] = sin(phi)*sqrt(1-rho*rho);
			r[3] = rho;
			theta = 2*asin(1)*(130/180.0);
			rot[1][1] = (1 - cos(theta))*r[1]*r[1] + cos(theta);
			rot[1][2] = (1 - cos(theta))*r[1]*r[2] - sin(theta)*r[3];
			rot[1][3] = (1 - cos(theta))*r[1]*r[3] + sin(theta)*r[2];
			rot[2][1] = (1 - cos(theta))*r[1]*r[2] + sin(theta)*r[3];
			rot[2][2] = (1 - cos(theta))*r[2]*r[2] + cos(theta);
			rot[2][3] = (1 - cos(theta))*r[2]*r[3] - sin(theta)*r[1];
			rot[3][1] = (1 - cos(theta))*r[1]*r[3] - sin(theta)*r[2];
			rot[3][2] = (1 - cos(theta))*r[2]*r[3] + sin(theta)*r[1];
			rot[3][3] = (1 - cos(theta))*r[3]*r[3] + cos(theta);
			for(int j = 1;j <= fluid_no[i]; j++) {
				k = cell_part[j][i];
				del_vx = vel_fl[3*k-2] - cell_velx[i];
				del_vy = vel_fl[3*k-1] - cell_vely[i];
				del_vz = vel_fl[3*k] - cell_velz[i];

				vel_fl[3*k-2] = cell_velx[i] + rot[1][1]*del_vx + rot[1][2]*del_vy + rot[1][3]*del_vz;
				vel_fl[3*k-1] = cell_vely[i] + rot[2][1]*del_vx + rot[2][2]*del_vy + rot[2][3]*del_vz;
				vel_fl[3*k]   = cell_velz[i] + rot[3][1]*del_vx + rot[3][2]*del_vy + rot[3][3]*del_vz;
			}
		}
	}
	for(int i = 1; i <= lx*ly*lz; i++) {
		var = 0.0;
		if(fluid_no[i] > 1) {
			for(int j =1; j <= fluid_no[i]; j++) {
				k = cell_part[j][i];
				del_vx1[k] = vel_fl[3*k - 2] - cell_velx[i];
				del_vy1[k] = vel_fl[3*k - 1] - cell_vely[i];
				del_vz1[k] = vel_fl[3*k]   - cell_velz[i];
				var += del_vx1[k]*del_vx1[k] + del_vy1[k]*del_vy1[k] + del_vz1[k]*del_vz1[k];
			}
			scale_fac_mpcd = sqrt(3.0 * (fluid_no[i] - 1) * kbt/(mass_fl * var));
			for(int j = 1; j <= fluid_no[i]; j++) {
				k = cell_part[j][i];
				vel_fl[3*k-2] = cell_velx[i] + del_vx1[k]*scale_fac_mpcd;
				vel_fl[3*k-1] = cell_vely[i] + del_vy1[k]*scale_fac_mpcd;
				vel_fl[3*k]   = cell_velz[i] + del_vz1[k]*scale_fac_mpcd;
			}
		}
	}
	for(int i = 0; i <= maxpart; i++)
		free(cell_part[i]);
	free(tmp_pos);
	free(cell_part);
	free(fluid_no);
}

File: fortran/rotation_mpcd.f90
subroutine rotation_mpcd      !rotation step for all fluid particles
	use all_parameters
	implicit none
    integer :: i,j,k,l1,l2,m
    integer :: fluid_no(lx*ly*lz),cell_part(maxpart,lx*ly*lz)
    double precision :: cell_velx(lx*ly*lz), cell_vely(lx*ly*lz),ct
    double precision :: r1, r2, r3, theta, ran1, phi, rho, prob, st
    double precision :: rot11,rot12,rot13,rot21,rot22,rot23,rot31
	double precision :: rot32,rot33,ir1, ir2, ir3
	double precision :: del_vx,del_vy,del_vz,mx1,my1,mz1, ict
	double precision :: tmp_pos(3*no_of_fluid), cell_velz(lx*ly*lz)
	double precision :: rr1,rr2,rr3,var,del_vx1(no_of_fluid),del_vy1(no_of_fluid),del_vz1(no_of_fluid)
	integer::cell_no

	fluid_no = 0

	rr1 = ran1(zzzz) - 0.5d0
	rr2 = ran1(zzzz) - 0.5d0
	rr3 = ran1(zzzz) - 0.5d0
	tmp_pos = pos_fl
	do i=1,no_of_fluid
    	tmp_pos(3*i-2)= tmp_pos(3*i-2) + rr1
        tmp_pos(3*i-1)= tmp_pos(3*i-1) + rr2
        tmp_pos(3*i)  = tmp_pos(3*i)   + rr3
        tmp_pos(3*i-2)= tmp_pos(3*i-2) - llx*anint((tmp_pos(3*i-2)-llxby2)*inv_llx)
        tmp_pos(3*i-1)= tmp_pos(3*i-1) - lly*anint((tmp_pos(3*i-1)-llyby2)*inv_lly)
        tmp_pos(3*i)= tmp_pos(3*i) - llz*anint((tmp_pos(3*i)-llzby2)*inv_llz)
    enddo

	do i=1,no_of_fluid
		cell_no = 1 + int(tmp_pos(3*i-2)) + lx*int(tmp_pos(3*i-1)) + lx*ly*int(tmp_pos(3*i))
		fluid_no(cell_no) = fluid_no(cell_no) + 1
        j = fluid_no(cell_no)
        cell_part(j,cell_no) = i
    end do

	cell_velx = 0.0d0; cell_vely = 0.0d0; cell_velz = 0.0d0

	do i=1,lx*ly*lz
    	if (fluid_no(i).gt.1) then
			do j=1,fluid_no(i)
           		k= cell_part(j,i)
           		cell_velx(i) = cell_velx(i) + vel_fl(3*k-2)/dfloat(fluid_no(i))
           		cell_vely(i) = cell_vely(i) + vel_fl(3*k-1)/dfloat(fluid_no(i))
           		cell_velz(i) = cell_velz(i) + vel_fl(3*k)/dfloat(fluid_no(i))
        	end do
        	rho = 2.0d0*ran1(zzzz)-1.0d0
        	phi = 4.0d0*asin(1.0d0)*ran1(zzzz)
        	r1 = cos(phi)*dsqrt(1-rho*rho)
        	r2 = sin(phi)*dsqrt(1-rho*rho)
        	r3 = rho
        	theta = 2.0d0*asin(1.0d0)*(130.0/180.0)    !!Why muliply by 2
			st = sin(theta)
			ct = cos(theta)
			ict = 1.0d0 - ct
			ir1 = ict*r1; ir2 = ict*r2; ir3 = ict*r3
        	rot11 = ir1*r1 + ct; rot12 = ir1*r2- st*r3; rot13 = ir1*r3 + st*r2
        	rot21 = ir2*r1 + st*r3; rot22 = ir2*r2 + ct; rot23 = ir2*r3 - st*r1
        	rot31 = ir3*r1 - st*r2; rot32 = ir3*r2 + st*r1; rot33 = ir3*r3 + ct
	    	do j=1,fluid_no(i)
           		k= cell_part(j,i)
           		del_vx = vel_fl(3*k-2) - cell_velx(i)
           		del_vy = vel_fl(3*k-1) - cell_vely(i)
           		del_vz = vel_fl(3*k)   - cell_velz(i)
		   		vel_fl(3*k-2) = rot11*del_vx + rot12*del_vy + rot13*del_vz
           		vel_fl(3*k-1) = rot21*del_vx + rot22*del_vy + rot23*del_vz
         	 	vel_fl(3*k)   = rot31*del_vx + rot32*del_vy + rot33*del_vz
		   		vel_fl(3*k-2) = cell_velx(i) + vel_fl(3*k-2)
		   		vel_fl(3*k-1) = cell_vely(i) + vel_fl(3*k-1)
           		vel_fl(3*k)   = cell_velz(i) + vel_fl(3*k)
			end do
		end if
	end do
    if (mod(nn,1)==0) then
		do i=1,lx*ly*lz
        	var = 0.0d0
         	if(fluid_no(i).gt.1) then
           		do j=1,fluid_no(i)
              		k=cell_part(j,i)
              		del_vx1(k) = vel_fl(3*k-2) - cell_velx(i)
              		del_vy1(k) = vel_fl(3*k-1) - cell_vely(i)
              		del_vz1(k) = vel_fl(3*k)   - cell_velz(i)
              		ict = del_vx1(k)**2 + del_vy1(k)**2 + del_vz1(k)**2
					var = var + ict
				end do
          		scale_fac_mpcd = dsqrt(3.0d0*(fluid_no(i)-1)*kbt/(mass_fl*var))
          		do j=1,fluid_no(i)
          			k=cell_part(j,i)
            		vel_fl(3*k-2) = cell_velx(i) + del_vx1(k)*scale_fac_mpcd
            		vel_fl(3*k-1) = cell_vely(i) + del_vy1(k)*scale_fac_mpcd
            		vel_fl(3*k)   = cell_velz(i) + del_vz1(k)*scale_fac_mpcd
         		end do
        	endif
		end do
	end if
end subroutine rotation_mpcd
